Notebook 1: Introduction to Statistics

1. What is Statistics?
   
Understand what statistics means and how we collect, organize, analyze and interpret data.

3. Why Statistics is Important in AI/ML
   
Understand how statistics helps with:

- Understanding datasets
- Finding patterns
- Data preprocessing
- Model evaluation
- Making decisions from data

3. Statistics vs Machine Learning
  
Understand the difference between:

- Statistics → understanding data and relationships
- Machine Learning → learning patterns from data to make predictions/decisions

4. Descriptive Statistics
Cover how we summarize existing data using things such as:
- Mean
- Median
- Mode
- Minimum
- Maximum
- Range
- Standard deviation
5. Inferential Statistics
Understand how we use a sample to draw conclusions about a larger population.
6. Population
The complete group we are interested in studying.
7. Sample
A smaller subset selected from the population.
8. Census
Collecting information from every member of the population.
9. Sampling
The process of selecting a sample from a population.
10. Variables
A characteristic or value that can change between observations.
11. Features
Input variables used by a machine-learning model to make predictions.
12. Observation
One individual record/data point in a dataset.
13. Dataset
A collection of observations and their variables/features.
14. Quantitative Data
Numerical data that can be measured or counted.
Examples:
- Age = 25
- Salary = ₹50,000
- Height = 176 cm
15. Qualitative Data
Descriptive or categorical data rather than numerical measurements.
Examples:
- Gender/category
- Product type
- City
- Customer feedback
16. Continuous Data
Numerical data that can take values within a range.
Examples:
- Height = 176.5 cm
- Temperature = 36.7°C
- Weight = 72.4 kg
17. Discrete Data
Countable numerical values, usually whole numbers.
Examples:
- Number of customers = 100
- Number of orders = 25
- Number of cars = 8
18. Structured Data
Data organized into a predefined structure, usually rows and columns.
Example:
Age	Salary	City
25	50000	Hyderabad
30	65000	Chennai


This is common in databases and CSV files.
19. Unstructured Data
Data without a fixed tabular structure.
Examples:
- Images
- Videos
- Audio
- Text documents
- Social-media posts
This is especially important in AI/ML, NLP, computer vision and Generative AI.

Notebook 2: Measures of Central Tendency


1. Mean
   
Step 1 – Understand the Concept

What is Mean?
Mean is the average value of a set of numbers.
We calculate it by adding all the values and dividing the total by the number of values.

Mean is useful when we want to represent a dataset with one central value.

For example, if five employees have salaries of ₹30,000, ₹35,000, ₹40,000, ₹45,000 and ₹50,000, the mean salary gives us the average salary of these employees.
Formula
\[
Mean = \frac{\sum x}{n}
\]
Where:
- \(x\) = each value
- \(n\) = number of values
- \(\sum x\) = sum of all values

AI/ML Example
Suppose a dataset contains customer ages:
20, 25, 30, 35, 40

The mean age can be used as a basic summary of the dataset and can help us understand the distribution of the data.

In [2]:
%pip install numpy


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.8/6.8 MB 13.0 MB/s  0:00:00 eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [3]:
import numpy as np

marks = np.array([60, 70, 80, 90, 100])

mean = np.mean(marks)

print("Mean:", mean)

Mean: 80.0


2. Median

Step 1 – Understand the Concept

What is Median?
Median is the middle value when the data is arranged in ascending or descending order.
If there is an even number of values, we take the average of the two middle values.

Median is especially useful when the dataset contains outliers because extreme values don't affect it as much as they affect the mean.

Formula

For an odd number of observations:
\[
Median = \text{middle value}
\]
For an even number:
\[
Median = \frac{\text{two middle values}}{2}
\]


In [4]:
import numpy as np

salaries = np.array([25000, 30000, 35000, 40000, 500000])

median = np.median(salaries)

print("Median:", median)

Median: 35000.0


3. Mode
   
Step 1 – Understand the Concept

What is Mode?
Mode is the value that occurs most frequently in a dataset.
Unlike mean and median, mode can be used directly with categorical data.

A dataset can have:
- One mode → unimodal
- Two modes → bimodal
- Multiple modes → multimodal
  
Formula
There isn't a mathematical calculation like mean.
The mode is simply:
\[
Mode = \text{most frequently occurring value}
\]

In [5]:
from collections import Counter

sizes = ["M", "L", "M", "S", "M", "L", "M"]

counts = Counter(sizes)

mode = counts.most_common(1)[0][0]

print("Mode:", mode)

Mode: M


Interpretation
M is the most frequently occurring shirt size.
Where is it used in AI/ML?
Mode can be useful for:
- Categorical data
- Missing categorical values
- Customer segmentation
- Survey analysis
- Understanding the most common category

4. Weighted Mean

Step 1 Understand the Concept

What is Weighted Mean?
Weighted mean is an average where some values have more importance than others.
In a normal mean, every value contributes equally.
    
In a weighted mean, each value has a weight.
    
Formula
\[
Weighted\ Mean = \frac{\sum wx}{\sum w}
\]
Where:
- \(x\) = value
- \(w\) = weight

Business Example

A company may calculate an employee's overall performance using:
- Project performance → 50%
- Technical skills → 30%
- Communication → 20%
  
The different weights represent different levels of importance.

AI/ML Example
Weighted averages can be used when combining metrics where some observations or categories should contribute more than others.

In [11]:
import numpy as np

scores = np.array([80, 70, 90])
weights = np.array([0.2, 0.3, 0.5])

weighted_mean = np.average(scores, weights=weights)

print("Weighted Mean:", weighted_mean)

Weighted Mean: 82.0


Interpretation

The final weighted score is 82.

The final exam has the highest weight, so its score has the largest effect on the result.
Where is it used in AI/ML?
Weighted means are useful in:
- Evaluation metrics
- Weighted datasets
- Business scoring systems
- Imbalanced datasets
- Combining multiple measurements

5. Geometric Mean
   
Step 1 – Understand the Concept

What is Geometric Mean?

Geometric mean is an average based on multiplication rather than addition.
It is particularly useful when working with:
- Growth rates
- Returns
- Ratios
- Multiplicative changes
  
Formula

For \(n\) values:
\[
Geometric\ Mean = \sqrt[n]{x_1x_2...x_n}
\]

In [12]:
import numpy as np

growth_factors = np.array([1.10, 1.20])

geometric_mean = np.prod(growth_factors) ** (1 / len(growth_factors))

print("Geometric Mean:", geometric_mean)
print("Average Growth Rate:", (geometric_mean - 1) * 100, "%")

Geometric Mean: 1.1489125293076057
Average Growth Rate: 14.891252930760569 %


Interpretation
The average compound growth factor is approximately 1.149, corresponding to about 14.9% growth.
Where is it used in AI/ML?
It can be useful when dealing with:
- Growth rates
- Ratios
- Multiplicative processes
- Certain performance metrics

6. Harmonic Mean
   
Step 1 – Understand the Concept

What is Harmonic Mean?

Harmonic mean is an average that is especially useful for rates and ratios.
It gives more importance to smaller values compared with the arithmetic mean.

Formula
\[
Harmonic\ Mean =
\frac{n}{\sum\frac{1}{x_i}}
\]
For two values:
\[
H = \frac{2ab}{a+b}
\]

Business Example
A company may use harmonic mean when combining performance expressed as rates, such as certain productivity or speed measurements.
AI/ML Example
One of the most important ML examples is the F1-score.
F1-score is the harmonic mean of:
- Precision
- Recall
\[
F1 = \frac{2PR}{P+R}
\]

In [13]:
import numpy as np

precision = 0.80
recall = 0.60

harmonic_mean = 2 * precision * recall / (precision + recall)

print("Harmonic Mean / F1 Score:", harmonic_mean)

Harmonic Mean / F1 Score: 0.6857142857142857


Interpretation

The F1-score is approximately 0.686.
It combines precision and recall into one value while giving importance to both.
Where is it used in AI/ML?
Harmonic mean is especially important for:
- Precision and Recall
- F1-score
- Classification problems
- Information retrieval
- Search systems

In [14]:
import os
print(os.path.abspath("notebook1.ipynb"))

/Users/harshzz/Sprint_2/notebook1.ipynb


In [15]:
import os

path = os.path.abspath("notebook1.ipynb")

print("JUPYTER FILE:")
print(path)

print("\nFILE SIZE:")
print(os.path.getsize(path), "bytes")

JUPYTER FILE:
/Users/harshzz/Sprint_2/notebook1.ipynb

FILE SIZE:
16688 bytes
